src/test/java/org/folio/auth/authtokenmodule/AuthTokenTest.java

```diff
@@ -1016,11 +1038,187 @@ public void testAdminHealth()  {
+        assertThat(e.getMessage(), containsString("revoked"));
+      }));
+  }
+
+  @Test
+  public void testStoreRefreshTokenSingleUse(TestContext context) {
+    var ts = new RefreshTokenStore(vertx, tenant);
+    // Create and save some tokens.
+    var rt1 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt2 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt3 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var s1 = ts.saveToken(rt1);
+    var s2 = ts.saveToken(rt2);
+    var s3 = ts.saveToken(rt3);
+
+    Async async = context.async();
+    CompositeFuture.all(s1, s2, s3)
+      .compose(a -> ts.checkTokenNotRevoked(rt2))
+      .onComplete(context.asyncAssertSuccess()) // First check should succeed.
+      .compose(a -> ts.checkTokenNotRevoked(rt2))
+      .onComplete(context.asyncAssertFailure(b ->
+        assertThat(b.getMessage(), containsString("revoked"))
+      ))
+      .compose(a -> ts.checkTokenNotRevoked(rt1))
+      .onComplete(context.asyncAssertFailure(b ->
+        assertThat(b.getMessage(), containsString("revoked"))
+      ))
+      .compose(a -> ts.checkTokenNotRevoked(rt3))
+      .onComplete(context.asyncAssertFailure(b -> {
+        assertThat(b.getMessage(), containsString("revoked"));
+        async.complete();
+      }));
+  }
+
+  @Test
+  public void testStoreCleanupExpired(TestContext context) {
+    // Create some tokens.
+    var rt1 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt2 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt3 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt4 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt5 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+
+    // Set a few tokens' expires at time to simulate expiration.
+    var now = Instant.now().getEpochSecond();
+    rt2.setExpiresAt(now - 10); // Would have expired 10 seconds ago.
+    rt4.setExpiresAt(now - 20); // Would have expired 20 seconds ago.
+
+    Async async = context.async();
+    var ts = new RefreshTokenStore(vertx, tenant);
+
+    // Other tests could have added tokens to storage, so remove all of those first.
+    // Then save 5 tokens, two of which are expired. When each token is saved the method
+    // cleans up expired tokens, so after all 5 have been saved, only 3 should be left.
+    ts.removeAll()
+      .compose(x -> {
+        var s1 = ts.saveToken(rt1);
+        var s2 = ts.saveToken(rt2);
+        var s3 = ts.saveToken(rt3);
+        var s4 = ts.saveToken(rt4);
+        var s5 = ts.saveToken(rt5);
+        return CompositeFuture.all(s1, s2, s3, s4, s5);
+      })
+      .compose(y -> ts.countTokensStored(tenant))
+      .onComplete(context.asyncAssertSuccess(count -> {
+        assertThat(count, is(3));
```

Comment:
The problem with this is that I call `cleanupExpiredTokens` in `onComplete` in `saveToken` to get the "fire and forget" behavior that we want. Because of this there is no guarantee here that `cleanupExpiredTokens` will complete in the test. To keep things simple I could test `cleanupExpiredTokens` separately (which I originally had) but I'll also need a `saveToken` method which doesn't clean up to do that. Is there a better way?

Reply:
```suggestion
  public Future<Void> saveToken(RefreshToken refreshToken) {
    return saveToken(refreshToken, Promise.promise());
  }
  
  Future<Void> saveToken(RefreshToken refreshToken, Promise<Void> promise) {
   ..
   
    return pool.preparedQuery(insert).execute(values)
      .onComplete(x -> cleanupExpiredTokens().onComplete(promise)).mapEmpty();
  }
  
```

Reply:
I'd use https://github.com/awaitility/awaitility

Reply:
Thanks for the ideas. Julian that looks like a cool library but I wasn't able to figure out how to get it to play nice with vertx futures in the time I gave myself. I went with something like what Adam has but used a boolean instead to control which method of waiting gets called. See what you think.